[Stats Revamp] Show data for a week in Views & Visitors and Total Likes details

staskus opened the pull request:
Fixes #19729

## Description

Details page cards for Views & Visitors - **Countries** and **Referrers** and details page cards for Total Likes - **Posts and pages** are showing data for a `day` and not for the `week`.

## Rootcause

The issue happens because we load `day` data for Views & Visitors and Total & Likes cards, that display daily changes in the diagrams. 

We call `refreshPeriodOverviewData` in `StatsPeriodStore` which refreshes the summary (for Views & Visitors / Total Likes) cards as well as all the other information in `fetchAsyncData` which is then used in `StatsInsightDetailsViewModel`. However, when refreshing we are using one period (`day` or `week`) which is used to fetch all the information.

## Solution

- After the discussion in the PR, decided to create a new `StatsRevampStore` as a beginning of store refactorings.
- Build separate methods: `refreshViewsAndVisitors` and `refreshLikesTotal`, which fetches summary for the `day` period and other required information for the `week` period. The solution is based fully on all the other methods built before in `StatsPeriodStore`.
- `StatsRevampStore` only exposes the data and actions for `viewsAndVisitors` and `likesTotals` to represent our necessary use-cases and hide complexities.

## Testing instructions

### Case 1 Views & Visitors - Countries & Referrers:
1. Open Jetpack
2. Enable "New appearance for Stats" & "New cards for Stats insights" in App Settings -> Debug
3. Go to stats
4. Click on settings icon on top right and add "Views & Visitors card"
5. Click "Week"
6. Confirm that "Views & Visitors", "Referrers", "Countries" cards appear. Compare the numbers with the Web (https://wordpress.com/stats/week/)
7. Change dates
8. Confirm that dates are changing and numbers match the web


### Case 2 Total likes - Posts and pages:
1. Open Jetpack
2. Enable "New appearance for Stats" & "New cards for Stats insights" in App Settings -> Debug
3. Go to stats
4. Click on settings icon on top right and add "Total likes"
5. Click "Week"
6. Confirm that "Total Likes, "Posts and pages" cards appear. Compare the numbers with the Web (https://wordpress.com/stats/week/$site)
7. Change dates
8. Confirm that dates are changing and numbers match the web

 **Known issue:**  [Stats: Total likes details view show information for the last 7 days instead of this week](https://github.com/wordpress-mobile/WordPress-iOS/issues/19781)

## Regression Notes

1. Potential unintended areas of impact

It should only affect `Views & Visitors` and `Total Likes` details

2. What I did to test those areas of impact (or what existing automated tests I relied on)

None

3. What automated tests I added (or what prevented me from doing so)

Added `StatsRevampStoreTests` 

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Images & Videos

## Before

Views & Visitors - Referrers and Countries show the wrong numbers

https://user-images.githubusercontent.com/4062343/206481739-fc83bf8a-4d0c-47c8-9c53-1a5e71eb07b4.mp4

## After

Views & Visitors - Referrers and Countries show the correct numbers

https://user-images.githubusercontent.com/4062343/206480497-62cd6c23-3cd8-4c17-aa2a-8d88b2837281.mp4

Total Likes - Post and Pages show the correct numbers

https://user-images.githubusercontent.com/4062343/207103504-5df54882-9224-4fec-9c76-fa9e42686987.mp4

staskus committed:
Add refreshViewsAndVisitors to StatsPeriodStore that loads summary, countries and referrers

Added ability to fetch summary for .day period and countries and referrers for .week period which was not supported for StatsPeriodStore. Using previously written code to refresh period overview as an example

staskus committed:
Use a new refreshViewsAndVisitors period action when loading insight details

staskus committed:
Allow reloading views and visitors details view with pull to refresh

staskus committed:
Add refreshTotalLikes to StatsPeriodStore that loads summary and topPostsAndPages

Added ability to fetch summary for .day period and post and pages for .week period which was not supported for StatsPeriodStore. Using previously written code to refresh period overview as an example

staskus committed:
Reusing fetchSummary code for both totalLikesDetails and viewsAndVisitorsDetails

staskus committed:
Do not show empty cards when total likes details is loading

staskus committed:
Merge branch 'trunk' into fix/stats-revamp-referrers-countries-posts-and-pages-show-incorrect-data

staskus committed:
Added tests asserting that correct states are changed when dispatching new refresh views and total likes actions

sla8c commented:
@staskus I found an issue on the TotalLikes count vs Web.  See screenshot

![image](https://user-images.githubusercontent.com/88816724/207532109-7dc8b69c-3912-44ef-8225-1c214b1705a9.png)

sla8c commented on WordPress/Classes/Stores/StatsPeriodStore.swift:
```diff
@@ -1473,3 +1489,208 @@ private extension StatsPeriodStore {
+    }
+
+    func setViewsAndVisitorsFetchingStatus(_ status: StoreFetchingStatus) {
+        if FeatureFlag.statsPerformanceImprovements.enabled {
```
Maybe we don't need the statsPerformanceImprovements flag anymore? The feature flag is enabled + production has been running with this enabled for awhile

staskus replied:
Yes, it does look like the flag could be removed.

sla8c commented on WordPress/Classes/Stores/StatsPeriodStore.swift:
```diff
@@ -1473,3 +1489,208 @@ private extension StatsPeriodStore {
+    }
+
+    func setTotalLikesDetailsFetchingStatus(_ status: StoreFetchingStatus) {
+        if FeatureFlag.statsPerformanceImprovements.enabled {
```
Maybe we don't need the statsPerformanceImprovements flag anymore? The feature flag is enabled + production has been running with this enabled for awhile

sla8c commented:
@staskus @guarani  I have been thinking about this one and I think that this might be the point where we start creating a separate Store class to specifically handle the Stats Revamp logic.  The code / functionality added to StatsPeriodStore starts to change the purpose of what the original dev's created StatsPeriodStore for which isn't a fault; this is just due to the new stats revamp screen designs blending different periods and or insights and data requirements onto the same screen(s).  I think the additional benefit of creating a new store sooner is StatsPeriodStore is already a key class for current live user functionality on Stats and separating into a diff class means we reduce the risk of a production bug impacting a majority of users.  WDYT?

staskus commented:
Thanks for the review, @sla8c ! I'll take a look why the numbers don't match the web 🤔

> @staskus @guarani I have been thinking about this one and I think that this might be the point where we start creating a separate Store class to specifically handle the Stats Revamp logic.

I agree that `StatsPeriodStore` is not the best place for this code, as I go against the idea of the class. 

I wonder what's the role of `StatsInsightsStore` vs `StatsPeriodStore`? Are they created to collect information from different endpoints? Because they are still both used in `SiteStatsInsightsViewModel`. 

I wonder how the new store should look and behave. In my idea:
- The interface of it could describe our required purposes. For now - detailsForViewsAndVisitors & detailsForTotalLikes
- The internals could either - a) reuse `StatsInsightsStore` and (multiple ?) `StatsPeriodStore` and just aggregate the required information. b) remake parts of required functionality, make API calls and do information storing itself.

What is your view? Am I thinking in the right direction from your POV?

sla8c commented:
@staskus `StatsInsightsStore` is similar to `StatsPeriodStore` and is a legacy Store for the current Insights tab / first screen on the production AppStore version.

What you described above is definitely along the lines that I am thinking.  Regarding class name though I'm not sure because as we know the Stats Revamp merges both Insights and Period data.  Maybe temporarily we could call it `StatsRevampStore` and we rename it properly later once we are able to decommission the old StatsStore classes?  I don't have a strong opinion on the name.

staskus commented:
Thanks for the answer!

> What you described above is definitely along the lines that I am thinking.

Good. I could try to make appropriate changes: Create a new store and leave existing stores unchanged.

> ...we could call it `StatsRevampStore` and we rename it properly later once we are able to decommission the old StatsStore classes? I don't have a strong opinion on the name.

Yes, I think that's reasonable.

staskus committed:
Merge branch 'trunk' into fix/stats-revamp-referrers-countries-posts-and-pages-show-incorrect-data

staskus committed:
Move insight details related code to a new StatsRevampStore

staskus committed:
Refactor StatsRevampStore to expose interface for 2 supported refresh methods

StatsRevampStore interface exposes statuses and data for 2 actions: viewsAndVisitors and likesTotals. It is done to hide complexities inside and aggregate data inside StatsRevampStore.

staskus committed:
Show total likes chart even if referrers not loaded in insights details

staskus committed:
Added a comment of StatsRevampStore

staskus committed:
Remove statsPerformanceImprovements feature flag

staskus committed:
Update logging text for StatsRevampStore

staskus commented:
> @staskus I found an issue on the TotalLikes count vs Web. See screenshot
> 
> ![image](https://user-images.githubusercontent.com/88816724/207532109-7dc8b69c-3912-44ef-8225-1c214b1705a9.png)

This is the same issue as [was fixed for Views & Visitors](https://github.com/wordpress-mobile/WordPress-iOS/issues/19345). I created a separate issue https://github.com/wordpress-mobile/WordPress-iOS/issues/19781

staskus commented:
@guarani, @sla8c  it's ready for testing. I made refactoring to isolate changes in a new class.

sla8c commented:
@staskus just fyi I've tested and functionality wise it works for me + I'm still in process of reviewing the code thx

sla8c commented on WordPress/Classes/Stores/StatsRevampStore.swift:
```diff
@@ -0,0 +1,373 @@
+                return
+        }
+
+        let summary = StatsRecord.timeIntervalData(for: blog, type: .blogVisitsSummary, period: StatsRecordPeriodType(remoteStatus: .day), date: date)
+        let referrers = StatsRecord.timeIntervalData(for: blog, type: .referrers, period: StatsRecordPeriodType(remoteStatus: .week), date: date)
+        let countries = StatsRecord.timeIntervalData(for: blog, type: .countryViews, period: StatsRecordPeriodType(remoteStatus: .week), date: date)
```
@staskus these calls here + loading data from cache are using other screens and datastore(s) such as `StatsPeriodStore` to save their data which is fine.  But I think if `StatsRevampStore` makes API calls it should also `persistToCoreData` as well so that its calls can be cached.

I did a test case using airplane mode where I did the following:
1. I went to Stats -> Insights -> Views and Visitors and tapped week.  I let the data load fully
2. I backed out and enabled airplane mode
3. On the Views and Visitors card I tapped week again.  This time I get a network error screen.  When I debugged this I can see the cached data for referrers and countries is nil.  I think summary data was cached because of the top level insights screen

An example of what I mean regarding `persistToCoreData` can be found here:
https://github.com/wordpress-mobile/WordPress-iOS/blob/d781476132c23c483065fde9aa28d3a59077056a/WordPress/Classes/Stores/StatsInsightsStore.swift#L197

sla8c commented on WordPress/Classes/Stores/StatsRevampStore.swift:
```diff
@@ -0,0 +1,373 @@
+                return
+        }
+
+        let summary = StatsRecord.timeIntervalData(for: blog, type: .blogVisitsSummary, period: StatsRecordPeriodType(remoteStatus: .day), date: date)
+        let posts = StatsRecord.timeIntervalData(for: blog, type: .topViewedPost, period: StatsRecordPeriodType(remoteStatus: .week), date: date)
```
I think the same cache issue will be here too?

sla8c commented:
@staskus this is looking pretty good; I left a couple of comments on the caching aspect for the new `StatsRevampStore`.  Thanks! 👍

staskus commented:
Thanks for the review, @sla8c! I will check the caching issues.

staskus committed:
Persisting data in StatsRevampStore after receiving from API

staskus commented:
@sla8c for now I am persisting data separately for each data type once it arrives. I do think this solution could be improved in the example of https://github.com/wordpress-mobile/WordPress-iOS/pull/19776 but I don't want to be duplicating the solutions.

sla8c reviewed (approved):
@staskus this looks now looks good to me.  Agree with your comment about incorporating the changes that I've done as part of https://github.com/wordpress-mobile/WordPress-iOS/pull/19776; maybe it makes sense to do in another PR?

Also I found an issue still with airplane mode showing the "Unable to load this content right now" message.  If the StatsRevampStore has the cached data I think the best user experience would be to display the cached data but this can be tackled in a different PR (if you'd prefer).  Thanks!

staskus committed:
Merge branch 'trunk' into fix/stats-revamp-referrers-countries-posts-and-pages-show-incorrect-data

staskus committed:
Do not show error state if cache is loaded for Views and Visitors and Total Likes details

staskus commented:
> @staskus this looks now looks good to me. Agree with your comment about incorporating the changes that I've done as part of #19776; maybe it makes sense to do in another PR?

Yes, I think it could be done separately.
 
> Also I found an issue still with airplane mode showing the "Unable to load this content right now" message. If the StatsRevampStore has the cached data I think the best user experience would be to display the cached data but this can be tackled in a different PR (if you'd prefer). Thanks!

I improved functionality for Views & Visitors and Total Like details but I noticed that this behaviour exists for all the Insights as well. 

I created an issue https://github.com/wordpress-mobile/WordPress-iOS/issues/19803
